### DISCUSSION TITLE
Changing compute offering / scaling VM and root disk

### DISCUSSION THREAD

**Discussion by vonlo123** [apache/cloudstack#8578] [Q&A] [answered]
We are having problems implementing our idea regarding the compute offerings.

We want to use offerings with a fixed amount of CPU, RAM and local storage so that the KVM host is utilized in the best possible way.
E.g. an offering 1:
1 CPU, 2 GB RAM, 20 GB local storage
Offering 2:
2 CPU 4 GB RAM, 40 GB local storage
etc.

We want to avoid a situation where a few VMs with little CPU/RAM consume the entire storage of the host so that the storage is fully booked but the CPU and RAM are not utilized. That would be a waste of resources.

Of course, we also want to offer the customer the option of changing the _compute offering_ if they need more resources for their VM.
According to the [documentation](https://docs.cloudstack.apache.org/en/latest/adminguide/storage.html#root-volume-size-defined-via-service-offering) it should be possible.

But after some testing, we have now come to the conclusion that changing an offering/scaling a VM is only possible if _Disk offering strictness_ is set to _false_ in the initial (and target) compute offering of the VM. 
But this would allow the user to freely adjust the size of the disk.

Is there a solution to this? Did we miss anything?
How do you ensure that all resources of a nodes are fully utilized?

Version 4.18.1.0 with KVM

**Comment by DaanHoogland** [comment]
@vonlo123 , did you investigate restricting the API for adjusting the diskoffering?
I expect you would want to disallow the use of https://cloudstack.apache.org/api/apidocs-4.18/apis/updateDiskOffering.html

I am not sure what you have configured in terms of roles, but expect you can work with those.

**Comment by weizhouapache** [comment]
@vonlo123 
there is a global setting `allow.diskoffering.change.during.scale.vm`
you can change the value to `true` and retry

**Reply by vonlo123** [reply]
Hello @weizhouapache 
we have changed the setting to `true` but the problem still remains. We can only choose a different compute offering when `disk offering strictness` is set to `false`

**Comment by vonlo123** [comment, marked as answer]
We solved this issue by denying _updateDIskOffering_
`disk offering strictness` needs to be set to `false`  

But this only works when the root disk in the compute offering has the _Provisioning type_ _fat_. With _thin provisioning_ I still can resize the disk. 
It looks like a bug...

**Reply by DaanHoogland** [reply]
Do you mean that the the user can still call `updateDiskOffering` despite it being denied in their role?
With thin provisioning the initial size should be just what is really used and it should auto grow up to it's limit.
Please describe the scenario you feel is symptomatic of a bug?

**Reply by NuxRo** [reply]
@DaanHoogland perhaps `resizeVolume` is the culprit here.
